### ISSUE TITLE
Deletion of nodes throws QueryExecutionException

### ISSUE THREAD

**Issue by InverseIntegral** [neo4j/neo4j#12878]
When I run the following two queries:

```
CREATE (:Q {p:false}), (:Q {p:false}), (:Q {p:true}), (:Q {p:false})
MATCH (n:Q) WHERE NOT ((n.p OR ("a"="")) XOR n.p) DELETE n
```

A `QueryExecutionException` is thrown:

```
Caused by: org.neo4j.graphdb.QueryExecutionException: Node with id 3 has been deleted in this transaction
	at org.neo4j.kernel.impl.query.QueryExecutionKernelException.asUserException(QueryExecutionKernelException.java:35)
	at org.neo4j.cypher.internal.javacompat.ResultSubscriber.converted(ResultSubscriber.java:426)
	at org.neo4j.cypher.internal.javacompat.ResultSubscriber.assertNoErrors(ResultSubscriber.java:407)
	at org.neo4j.cypher.internal.javacompat.ResultSubscriber.init(ResultSubscriber.java:85)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:116)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:285)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:274)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:268)
	... 3 more
Caused by: org.neo4j.kernel.impl.query.QueryExecutionKernelException: Node with id 3 has been deleted in this transaction
	... 12 more
Caused by: org.neo4j.exceptions.EntityNotFoundException: Node with id 3 has been deleted in this transaction
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundReadQueryContext.getTxStateNodePropertyOrNull(TransactionBoundQueryContext.scala:877)
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundReadQueryContext$NodeReadOperations.getTxStateProperty(TransactionBoundQueryContext.scala:916)
	at org.neo4j.cypher.internal.planning.ExceptionTranslatingReadQueryContext$ExceptionTranslatingReadOperations.getTxStateProperty(ExceptionTranslatingQueryContext.scala:378)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.AbstractCachedNodeProperty.getTxStateProperty(CachedNodeProperty.scala:84)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.AbstractCachedProperty.apply(CachedNodeProperty.scala:59)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.predicates.CoercedPredicate.isMatch(Predicate.scala:531)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.predicates.Predicate.apply(Predicate.scala:53)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.predicates.Predicate.apply(Predicate.scala:51)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.FilterPipe.$anonfun$internalCreateResults$1(FilterPipe.scala:32)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.FilterPipe.$anonfun$internalCreateResults$1$adapted(FilterPipe.scala:32)
	at org.neo4j.cypher.internal.runtime.ClosingIterator$$anon$2.innerHasNext(ClosingIterator.scala:140)
	at org.neo4j.cypher.internal.runtime.ClosingIterator.hasNext(ClosingIterator.scala:93)
	at org.neo4j.cypher.internal.runtime.ClosingIterator$$anon$5.innerHasNext(ClosingIterator.scala:191)
	at org.neo4j.cypher.internal.runtime.ClosingIterator.hasNext(ClosingIterator.scala:93)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.DistinctPipe$$anon$1.produceNext(DistinctPipe.scala:48)
	at org.neo4j.cypher.internal.runtime.PrefetchingIterator.pullNextElementFromSource(PrefetchingIterator.scala:44)
	at org.neo4j.cypher.internal.runtime.PrefetchingIterator.next(PrefetchingIterator.scala:38)
	at org.neo4j.cypher.internal.runtime.ClosingIterator$$anon$2.innerHasNext(ClosingIterator.scala:139)
	at org.neo4j.cypher.internal.runtime.ClosingIterator.hasNext(ClosingIterator.scala:93)
	at org.neo4j.cypher.internal.runtime.ClosingIterator$$anon$3.innerHasNext(ClosingIterator.scala:152)
	at org.neo4j.cypher.internal.runtime.ClosingIterator.hasNext(ClosingIterator.scala:93)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EmptyResultPipe.internalCreateResults(EmptyResultPipe.scala:29)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:82)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:78)
	at org.neo4j.cypher.internal.runtime.interpreted.PipeExecutionResult.request(PipeExecutionResult.scala:69)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.request(StandardInternalExecutionResult.scala:90)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.initiate(StandardInternalExecutionResult.scala:53)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.$anonfun$initiate$1(ClosingExecutionResult.scala:58)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.safely(ClosingExecutionResult.scala:99)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.initiate(ClosingExecutionResult.scala:58)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult$.wrapAndInitiate(ClosingExecutionResult.scala:183)
	at org.neo4j.cypher.internal.CypherCurrentCompiler$CypherExecutableQuery.innerExecute(CypherCurrentCompiler.scala:442)
	at org.neo4j.cypher.internal.CypherCurrentCompiler$CypherExecutableQuery.execute(CypherCurrentCompiler.scala:379)
	at org.neo4j.cypher.internal.ExecutionEngine.doExecute(ExecutionEngine.scala:217)
	at org.neo4j.cypher.internal.ExecutionEngine.$anonfun$executeSubquery$1(ExecutionEngine.scala:178)
	at org.neo4j.cypher.internal.ExecutionEngine.closing(ExecutionEngine.scala:183)
	at org.neo4j.cypher.internal.ExecutionEngine.executeSubquery(ExecutionEngine.scala:175)
	at org.neo4j.cypher.internal.ExecutionEngine.execute(ExecutionEngine.scala:125)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:127)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:115)
	... 8 more
```

**Neo4j Version:** 4.4.6
**Operating System:** 5.17.3 Arch Linux
**API:** Java

### Expected behavior
I would expect the queries not to thrown any exception.

### Actual behavior
An exception is thrown when executing the second query.

**Comment by HannesSandberg** [comment]
Thank you for the report! We will look into this and come back to you.

**Comment by sherfert** [comment]
This has been fixed in 4.3.16, 4.4.9 and 5.0.0
